smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,464 @@
+
+        self.logger = logging.getLogger(
+            self.__module__ + "." + self.__class__.__name__)
+
+        # INSTANCES
+
+        self.n_seeds = n_seeds if n_seeds else 1
+        self.instance_order = instance_order
+
+        # if instances are coming from Hyperband, skip the instance preprocessing section
+        # it is already taken care by Hyperband
+        if instances is not None and isinstance(instances[0], tuple):
+            self.instances = list(instances)
+        else:
+            instances = [] if instances is None else instances
+            # removing duplicates in the user provided instances
+            instances = list(OrderedDict.fromkeys(instances))
+
+            # determine instance order
+            if self.instance_order == 'shuffle_once':
+                # randomize once
+                self.rs.shuffle(instances)
+
+            # set seed(s) for all SH runs
+            # - currently user gives the number of seeds to consider
+            if self.deterministic:
+                seeds = [0]
+            else:
+                seeds = self.rs.randint(low=0, high=MAXINT, size=self.n_seeds)
+            # storing instances & seeds as tuples
+            self.instances = [(i, s) for s in seeds for i in instances]
+
+        # successive halving parameters
+        self._init_sh_params(initial_budget, max_budget, eta, num_initial_challengers)
+
+        # adaptive capping
+        if not self.cutoff_as_budget and self.instance_order != 'shuffle' and self.run_obj_time:
+            self.adaptive_capping = True
+        else:
+            self.adaptive_capping = False
+
+    def _init_sh_params(self, initial_budget: typing.Optional[float],
+                        max_budget: typing.Optional[float],
+                        eta: float,
+                        num_initial_challengers: typing.Optional[int]) -> None:
+        """
+        initialize Successive Halving parameters
+
+        Parameters
+        ----------
+        initial_budget : typing.Optional[float]
+            minimum budget allowed for 1 run of successive halving
+        max_budget : typing.Optional[float]
+            maximum budget allowed for 1 run of successive halving
+        eta : float
+            'halving' factor after each iteration in a successive halving run
+        num_initial_challengers : typing.Optional[int]
+            number of challengers to consider for the initial budget
+        """
+
+        if eta <= 1:
+            raise ValueError('eta must be greater than 1')
+        self.eta = eta
+
+        # BUDGETS
+
+        if max_budget is not None and initial_budget is not None \
+                and max_budget < initial_budget:
+            raise ValueError('Max budget has to be larger than min budget')
+
+        # - if only 1 instance was provided & quality objective, then use cutoff as budget
+        # - else, use instances as budget
+        if not self.run_obj_time and len(self.instances) <= 1:
```

Comment:
I believe there is an issue here. If the user wants to optimize a non-deterministic algorithm for multiple budgets, this is accidentally sorted into the multi-instance case. I actually don't know what to do in this situation. @mlindauer any suggestions? Until we have a solution for this we should maybe simply catch this case and raise an exception.

Reply:
This isn't resolved yet. Could you please add an `elif` here and fail the optimization?

Reply:
Adding that as an `elif` was not feasible because the instances would have already been processed by then. I check for this case in `__init__`. I have also added a "NOTE" for reference.